File: candy-007.js
// 編號：CANDY-007
// 程式語言：JavaScript
// 題目：在某個數字陣列裡，可能藏有某個不合群的奇數或偶數，試著找出它！

function findSomeDifferent(numbers) {
  let oddElements = numbers.filter((el) => el % 2 == 1);
  let evenElements = numbers.filter((el) => el % 2 == 0);
  return Number(oddElements.length > 1 ? evenElements : oddElements);
}

console.log(findSomeDifferent([2, 4, 0, 100, 4, 11, 2602, 36])); // 印出 11
console.log(findSomeDifferent([160, 3, 1719, 19, 11, 13, -21])); // 印出 160

// solving steps:
// 宣告兩個變數各自存有奇數與偶數元素的陣列，並以陣列長度判斷，如果奇數陣列的長度有超過一個，表示這陣列大部分是奇數，那我就要印出不是奇數的那個數值，反之亦然。

// CANDY-007 --- revision 針對誤用 Number() 做修正！
function findSomeDifferent(numbers) {
  let oddElements = numbers.filter((el) => el % 2 == 1);
  let evenElements = numbers.filter((el) => el % 2 == 0);
  let parseResult = oddElements.length > 1 ? evenElements : oddElements;
  return parseResult[0];
}

console.log(findSomeDifferent([2, 4, 0, 100, 4, 11, 2602, 36])); // 印出 11
console.log(findSomeDifferent([160, 3, 1719, 19, 11, 13, -21])); // 印出 160

// 編號：CANDY-007 --- revision01（不要取多餘的代名詞）
function findSomeDifferent(numbers) {
  let oddElements = numbers.filter((el) => el % 2 == 1);
  let evenElements = numbers.filter((el) => el % 2 == 0);
  return (oddElements.length > 1 ? evenElements : oddElements)[0];
}

console.log(findSomeDifferent([2, 4, 0, 100, 4, 11, 2602, 36])); // 印出 11
console.log(findSomeDifferent([160, 3, 1719, 19, 11, 13, -21])); // 印出 160

File: CandyCrush.rb/candy-007.rb
# # 編號：CANDY-007
# # 程式語言：Ruby
# # 題目：在某個數字陣列裡，可能藏有某個不合群的奇數或偶數，試著找出它！

# def findSomeDifferent(numbers)
#   even = numbers.select {|element| element % 2 == 0}
#   odd = numbers.select {|element| element % 2 == 1}
#   even.length > odd.length ? odd : even
# end

# puts findSomeDifferent([2, 4, 0, 100, 4, 11, 2602, 36]); # 印出 11
# puts findSomeDifferent([160, 3, 1719, 19, 11, 13, -21]); #印出 160

# # solving steps:
# # 用 select 方法篩選出偶數、奇數陣列，再判斷如果偶數陣列長度大於奇數的，則回傳奇數，反之亦然。

# 編號：CANDY-007 -- revision 將回傳型別修正為數字！
# 程式語言：Ruby
# 題目：在某個數字陣列裡，可能藏有某個不合群的奇數或偶數，試著找出它！

def findSomeDifferent(numbers)
  even = numbers.select {|element| element % 2 == 0}
  odd = numbers.select {|element| element % 2 == 1}
  even.length > odd.length ? odd.first : even.first
end

puts findSomeDifferent([2, 4, 0, 100, 4, 11, 2602, 36]); # 印出 11
puts findSomeDifferent([160, 3, 1719, 19, 11, 13, -21]); #印出 160

# solving steps:
# 用 select 方法篩選出偶數、奇數陣列，再判斷如果偶數陣列長度大於奇數的，則回傳奇數，反之亦然。